qiita_db/util.py

```diff
@@ -2673,3 +2678,216 @@ def _resource_allocation_failures(df, k, a, b, model, col_name, type_):
     df[f'c{type_}'] = model(x_plot, k, a, b)
     failures_df = df[df[type_] > df[f'c{type_}']]
     return failures_df
+
+
+def MaxRSS_helper(x):
+    if x[-1] == 'K':
+        y = float(x[:-1]) * 1000
+    elif x[-1] == 'M':
+        y = float(x[:-1]) * 1000000
+    elif x[-1] == 'G':
+        y = float(x[:-1]) * 1000000000
+    else:
+        y = float(x)
+    return y
+
+
+def update_resource_allocation_table(test=None):
+    # Thu, Apr 27, 2023 old allocations (from barnacle) were changed to a
+    # better allocation so using job 1265533 as the before/after so we only
+    # use the latests for the newest version
+    df = pd.DataFrame()
+
+    sql_command = """
+            SELECT
+                pj.processing_job_id AS processing_job_id
+            FROM
+                qiita.software_command sc
+            JOIN
+                qiita.processing_job pj ON pj.command_id = sc.command_id
+            JOIN
+                qiita.processing_job_status pjs
+                ON pj.processing_job_status_id = pjs.processing_job_status_id
+            LEFT JOIN
+                qiita.slurm_resource_allocations sra
+                ON pj.processing_job_id = sra.processing_job_id
+            WHERE
+                pjs.processing_job_status = 'success'
+            AND
+                pj.external_job_id ~ '^[0-9]+$'
+            AND
+                pj.external_job_id::INT >= 1265533
+            AND
+                sra.processing_job_id IS NULL;
+        """ if test is None else """
+        SELECT
+            pj.processing_job_id AS processing_job_id
+        FROM
+            qiita.software_command sc
+        JOIN
+            qiita.processing_job pj ON pj.command_id = sc.command_id
+        JOIN
+            qiita.processing_job_status pjs
+            ON pj.processing_job_status_id = pjs.processing_job_status_id
+        LEFT JOIN
+            qiita.slurm_resource_allocations sra
+            ON pj.processing_job_id = sra.processing_job_id
+        WHERE
+            pjs.processing_job_status = 'success'
+        AND
+            sra.processing_job_id IS NULL;
+    """
+
+    with qdb.sql_connection.TRN:
+        sql = sql_command
+        qdb.sql_connection.TRN.add(sql)
+        res = qdb.sql_connection.TRN.execute_fetchindex()
+        columns = ["processing_job_id"]
+        df = pd.DataFrame(res, columns=columns)
+
+    data = []
+    sacct = ['sacct', '-p', '--format=JobID,ElapsedRaw,MaxRSS,Submit,Start,'
+             'CPUTimeRAW,ReqMem,AllocCPUs,AveVMSize', '-j']
```

Comment:
Rather than what could be a very large number of calls to `sacct`, why not batch jobs and query `','.join(list_of_job_ids)`? The primary concern is creating a high burden for the resource manager, and secondary is the overhead of system calls

Reply:
I actually discussed this with @Gossty and this decision was made based on my previous experience with `sacct` and `pgsql`. As far as I was able to tell, `sacct` didn't create any burden to the resource manager - I think is because it is an actual database, which is not actually directly connected to the resource manager. Then in the past we have had issues with large inserts (vs. multiple single inserts) so I think basic inserts are the way to go here.

Reply:
See the warning [here](https://jhpce.jhu.edu/slurm/tips-sacct/)

Reply:
"That database needs to be modified constantly as jobs start and complete, so we don't want it tied up answering sacct queries." 

As you know, SQL engines are really good at bulk queries as there is a lot of overhead for a single query.